### PR TITLE
🥅 Add alternative NFT related routes

### PR DESCRIPTION
**Pull request by nwingt**
#  Test Cases
Now support the following routes

## ISCN ID
- [/?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)
- [/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)
- [/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)
- [/iscn/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/iscn/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)
- [/in/like/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/in/like/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)
- [/in/embed/nft/iscn/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1](https://button.rinkeby.like.co/in/embed/nft/iscn/iscn%3A%2F%2Flikecoin-chain%2FROQTh0xtbsKLZ-w9AYPrVbD2zWIpsFnpIyGaGDHU0mY%2F1)

## ISCN ID (Minted NFT)
- [/?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)
- [/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)
- [/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)
- [/iscn/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/iscn/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)
- [/in/like/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/in/like/iscn?iscn_id=iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)
- [/in/embed/nft/iscn/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1](https://button.rinkeby.like.co/in/embed/nft/iscn/iscn%3A%2F%2Flikecoin-chain%2FZzLfyk58twt4c6-ajNQi8rGZpdrEqPO_dREoYixud44%2F1)

## NFT Class ID
- [/?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)
- [/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)
- [/nft?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/nft?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)
- [/nft/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/nft/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)
- [/in/like/nft?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/in/like/nft?class_id=likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)
- [/in/embed/nft/class/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg](https://button.rinkeby.like.co/in/embed/nft/class/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)

### CODE REVIEW

**Comment by williamchong** [comment]
we also need to modify the oembed regex before this change deploy

File: src/util/nft.js [hunk 1]
```diff
@@ -25,3 +25,7 @@ export async function getClassInfo(classId) {
   if (nftClass) nftClass = parseNFTClassDataFields(nftClass);
   return nftClass;
 }
+
+export const checkIsValidISCNId = value => value.startsWith('iscn://likecoin-chain');
+
+export const checkIsValidNFTClassId = value => value.startsWith('likenft1');
```
Comment (williamchong) [inline src/util/nft.js:31]: We do have some regex we can use, but I guess these would do

**Review by williamchong — APPROVED** [review]

**Comment by nwingt** [comment]
> we also need to modify [the oembed regex](https://github.com/likecoin/likecoin-api-public/pull/341) before this change deploy

You mean adding `/in/nft/(:classId|:iscnId)`?

https://github.com/likecoin/likecoin-api-public/blob/57d443de8f250cec3911dcad5f67ea891c31ef58/src/routes/oembed/index.js#L80-L90

https://github.com/likecoin/likecoin-api-public/blob/57d443de8f250cec3911dcad5f67ea891c31ef58/src/routes/oembed/index.js#L130

**Comment by nwingt** [comment]
Case 2 not working due to issue on Firebase Function

**Comment by williamchong** [comment]
> > we also need to modify [the oembed regex](https://github.com/likecoin/likecoin-api-public/pull/341) before this change deploy
> 
> You mean adding `/in/nft/(:classId|:iscnId)`?
> 
> https://github.com/likecoin/likecoin-api-public/blob/57d443de8f250cec3911dcad5f67ea891c31ef58/src/routes/oembed/index.js#L80-L90
> 
> https://github.com/likecoin/likecoin-api-public/blob/57d443de8f250cec3911dcad5f67ea891c31ef58/src/routes/oembed/index.js#L130

i mean the /in/like regex in api is actually due to hardcode redirect, in which this PR will fix
so we should change the oembed api regex to use the new redirected path for iscn and nft, instead of/in/like

**Comment by nwingt** [comment]
The ISCN ID or the NFT Class ID is in the qs of the updated redirect path, not in the path, so it should be working without changing the API RegEx?

File: src/pages/in/nft/_id.vue [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        params: { id: 'iscn' },
+        query: { ...query, iscn_id: id }
+      });
+    } else if (checkIsValidNFTClassId(id)) {
+      redirect({
+        name: 'in-nft',
+        query: { ...query, class_id: id },
+      });
+    } else {
+      error({ statusCode: 400, message: 'INVALID_NFT_IDENTIFIER' });
```
Comment (williamchong) [inline src/pages/in/nft/_id.vue:23]: we can actually do [optional params in route](https://nuxtjs.org/docs/features/file-system-routing/#dynamic-routes) to merge this file with `index?

File: src/pages/iscn/_id.vue [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  fetch({ redirect, params, query }) {
+    // Redirect
+    // /iscn/:iscn_id  -> /in/like/iscn?iscn_id=:iscn_id
+    redirect({
+      name: 'in-like-id',
+      params: { id: 'iscn' },
+      query: { iscn_id: params.id, ...query },
+    });
+  },
+};
```
Comment (williamchong) [inline src/pages/iscn/_id.vue:12]: we can actually do [optional params in route](https://nuxtjs.org/docs/features/file-system-routing/#dynamic-routes) to merge this file with `index?

File: src/pages/nft/_id.vue [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    if (checkIsValidISCNId(id)) {
+      // /nft/:iscn_id  -> /in/like/iscn?iscn_id=:iscn_id
+      redirect({
+        name: 'in-like-id',
+        params: { id: 'iscn' },
+        query: { iscn_id: id, ...query },
+      });
+    } else if (checkIsValidNFTClassId(id)) {
+      // /nft/:class_id -> /in/nft?class_id=:class_id
+      redirect({ name: 'in-nft', query: { class_id: id, ...query } });
```
Comment (williamchong) [inline src/pages/nft/_id.vue:21]: we can actually do [optional params in route](https://nuxtjs.org/docs/features/file-system-routing/#dynamic-routes) to merge this file with `index?

**Review by williamchong — CHANGES_REQUESTED** [review]
Do we also want to handle unencoded iscn id via error page?
Smilar to https://github.com/likecoin/liker-land/pull/527

**Review by williamchong — APPROVED** [review]

**Pull request merged**